File: 백준/Gold/2533. 사회망 서비스（SNS）/사회망 서비스（SNS）.py
import sys
input = sys.stdin.readline

sys.setrecursionlimit(10**8)
def dfs(num):
    visited[num] = True
    dp[num][0] = 1
    for i in graph[num]:
        if not visited[i]:
            dfs(i)
            dp[num][1] += dp[i][0]
            dp[num][0] += min(dp[i][0],dp[i][1])



n =  int(input())
graph = [[] for _ in range(n+1)]

for _ in range(n-1):
    v1, v2 = map(int,sys.stdin.readline().split())
    graph[v1].append(v2)
    graph[v2].append(v1)

dp = [[0,0] for _ in range(n+1)]
visited = [False]* (n+1)

dfs(1)

print(min(dp[1][0], dp[1][1]))

File: 백준/Gold/2533. 사회망 서비스（SNS）/사회망 서비스（SNS）.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static int n;
    static int[][] dp;
    static ArrayList<Integer>[] graph;
    static boolean[] visited;
    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        n = Integer.parseInt(br.readLine());
        dp = new int[n][2];
        
        graph = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
        }
        for (int i = 0; i < n - 1; i++) {
            st = new StringTokenizer(br.readLine());
            int n1 = Integer.parseInt(st.nextToken()) - 1;
            int n2 = Integer.parseInt(st.nextToken()) - 1;
            graph[n1].add(n2);
            graph[n2].add(n1);
        }
        visited = new boolean[n];
        visited[0] = true;

        dfs(0);
        System.out.println(Math.min(dp[0][0], dp[0][1]));
    }

    private static void dfs(int root) {
        int minZero = 0;
        int minOne = 0;
        for (int a : graph[root]) {
            if (!visited[a]) {
                visited[a] = true;
                dfs(a);
            }
            minZero += Math.min(dp[a][0], dp[a][1]);
            minOne += dp[a][0];
        }
        dp[root][0] = minZero + 1;
        dp[root][1] = minOne;


    }
}